chromium_src/third_party/blink/renderer/modules/storage/dom_window_storage.cc

```diff
@@ -31,25 +33,213 @@ void MaybeClearAccessDeniedException(StorageArea* storage,
 
 }  // namespace
 
-BraveDOMWindowStorage::BraveDOMWindowStorage() {}
-BraveDOMWindowStorage::~BraveDOMWindowStorage() {}
+class EphemeralStorageNamespaces
+    : public GarbageCollected<EphemeralStorageNamespaces>,
+      public Supplement<Page> {
+ public:
+  EphemeralStorageNamespaces(StorageController* controller,
+                             const String& session_storage_id,
+                             const String& local_storage_id);
+  virtual ~EphemeralStorageNamespaces() = default;
+
+  static const char kSupplementName[];
+  static EphemeralStorageNamespaces* From(Page* page, LocalDOMWindow* window);
+
+  StorageNamespace* session_storage() { return session_storage_.Get(); }
+  StorageNamespace* local_storage() { return local_storage_.Get(); }
+  void Trace(Visitor* visitor) const override;
+
+ private:
+  Member<StorageNamespace> session_storage_;
+  Member<StorageNamespace> local_storage_;
+};
+
+const char EphemeralStorageNamespaces::kSupplementName[] =
+    "EphemeralStorageNamespaces";
+
+EphemeralStorageNamespaces::EphemeralStorageNamespaces(
+    StorageController* controller,
+    const String& session_storage_id,
+    const String& local_storage_id)
+    : session_storage_(
+          MakeGarbageCollected<StorageNamespace>(controller,
+                                                 session_storage_id)),
+      local_storage_(MakeGarbageCollected<StorageNamespace>(controller,
+                                                            local_storage_id)) {
+}
+
+void EphemeralStorageNamespaces::Trace(Visitor* visitor) const {
+  visitor->Trace(session_storage_);
+  visitor->Trace(local_storage_);
+  Supplement<Page>::Trace(visitor);
+}
 
 // static
-StorageArea* BraveDOMWindowStorage::sessionStorage(LocalDOMWindow& window,
-                                              ExceptionState& exception_state) {
-  auto* storage =
-      DOMWindowStorage::From(window).sessionStorage(exception_state);
-  MaybeClearAccessDeniedException(storage, window, &exception_state);
-  return storage;
+EphemeralStorageNamespaces* EphemeralStorageNamespaces::From(
+    Page* page,
+    LocalDOMWindow* window) {
+  DCHECK(window);
+  if (!page)
+    return nullptr;
+
+  EphemeralStorageNamespaces* supplement =
+      Supplement<Page>::From<EphemeralStorageNamespaces>(page);
+  if (supplement)
+    return supplement;
+
+  Document* document = window->GetFrame()->GetDocument();
+  auto* web_frame = WebLocalFrameImpl::FromFrame(document->GetFrame());
+  WebViewImpl* webview = web_frame->ViewImpl();
+  if (!webview || !webview->Client())
+    return nullptr;
+  String session_storage_id =
+      String::FromUTF8(webview->Client()->GetSessionStorageNamespaceId()) +
+      String("/ephemeral-session-storage");
+
+  auto* security_origin =
+      page->MainFrame()->GetSecurityContext()->GetSecurityOrigin();
+  String domain = security_origin->RegistrableDomain();
```

Comment:
just curious whether the protocol should participate in the id? I gather we should have different storages for http and https, does it work like this?

Reply:
It isn't working like that right now and, in fact, subdomains are also lost here. @bridiver, can you confirm is that's what you expected to happen here?